### PR TITLE
gh-117584: Raise TypeError for non-paths in `posixpath.relpath()`

### PR DESCRIPTION
**Pull request by nineteendo**
* Issue: gh-117584

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2024-04-06-16-42-34.gh-issue-117584.hqk9Hn.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Raise TypeError for non-paths in :func:`posixpath.relpath()`.
```
Comment (aisk) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-06-16-42-34.gh-issue-117584.hqk9Hn.rst]:
```suggestion
Raise :exec:`TypeError` for non-paths in :func:`posixpath.relpath()`.
```

**Comment by aisk** [comment]
If the `ValueError` is raised by `if not path:`, the falsy values `0` and `[]` will also raise the same `ValueError`, which is also strange. Maybe we can change the check to `if path == ""`.

File: Misc/NEWS.d/next/Core and Builtins/2024-04-06-16-42-34.gh-issue-117584.hqk9Hn.rst (hunk 1 above)
Comment (nineteendo) [reply Misc/NEWS.d/next/Core and Builtins/2024-04-06-16-42-34.gh-issue-117584.hqk9Hn.rst]: Should be `:exc:`.

**Comment by nineteendo** [comment]
No we can not, that doesn't work for Pathlike objects nor bytes.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by nineteendo** [comment]
@serhiy-storchaka can we back port this to 3.12? It's a pure bug fix.

**Comment by serhiy-storchaka** [comment]
I do not think this is a bug fix. Normally you should not pass `None` to `relpath()`, so this does not affect most of applications. If for some reasons you passed an argument of wrong type without checking it, and handle the resulting ValueError, changing the exception type to TypeError is a breaking change. Even if it is right in the long run, it should not be done is the bugfix release.

**Comment by nineteendo** [comment]
Thanks, closed the pull request.

**Pull request merged**